### PR TITLE
Add decoders for balancer swaps and mints/burns

### PR DESCRIPTION
**Pull request by yabirgb**
Closes #(issue_number)

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+V2_SWAP = b'!p\xc7A\xc4\x151\xae\xc2\x0e|\x10|$\xee\xcf\xdd\x15\xe6\x9c\x9b\xb0\xa8\xdd7\xb1\x84\x0b\x9e\x0b {'  # noqa: E501
+JOIN_V1 = b'c\x98-\xf1\x0e\xfd\x8d\xfa\xaa\xa0\xfc\xc7\xf5\x0b-\x93\xb7\xcb\xa2l\xccH\xad\xee(s"\rH]\xc3\x9a'  # noqa: E501
+EXIT_V1 = b'\xe7L\x91U+d\xc2\xe2\xe7\xbd%V9\xe0\x04\xe6\x93\xbd>\x1d\x01\xcc3\xe6V\x10\xb8j\xfc\xc1\xff\xed'  # noqa: E501
+VAULT_ADDRESS = string_to_evm_address('0xBA12222222228d8Ba445958a75a0704d566BF2C8')
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class BalancerDecoder(DecoderInterface):  # lgtm[py/missing-call-to-init]
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:37]: This pyling warning is not needed

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+        ),
+    ]
+
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x7716a99194d758c8537F056825b75Dd0C8FDD89f']])  # noqa: E501
+def test_v1_exit(database, ethereum_manager, eth_transactions):
+    """
+    Data for deposit taken from
+    https://etherscan.io/tx/0xfa1dfeb83480e51a15137a93cb0eba9ac92c1b6b0ee0bd8551a422c1ed83695b
+    Deposit 2.1 ether and borrow 4752 LUSD
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:277]: This comments were not updated.  Is a copy/paste

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    ]
+
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x7716a99194d758c8537F056825b75Dd0C8FDD89f']])  # noqa: E501
+def test_v1_join(database, ethereum_manager, eth_transactions):
+    """
+    Data for deposit taken from
+    https://etherscan.io/tx/0xb9dff9df4e3838c75d354d62c4596d94e5eb8904e07cee07a3b7ffa611c05544
+    Deposit 2.1 ether and borrow 4752 LUSD
+    """
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:146]: This comment is not updated. It was a copy/paste

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+from rotkehlchen.utils.hexbytes import hexstring_to_bytes
+
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x20A1CF262Cd3A42a50D226fD728104119e6fD0a1']])  # noqa: E501
+def test_v2_swap(database, ethereum_manager, eth_transactions):
+    """
+    Data for deposit taken from
+    https://etherscan.io/tx/0x35dd639ba80940cb14d79c965002a11ea2aef17bbf1f1b85cc03c336da1ddebe
+    Deposit 2.1 ether and borrow 4752 LUSD
+    """
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:24]: This comment was not updated. It was a copy/paste

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+from rotkehlchen.constants.assets import A_BAL, A_BPT, A_DAI, A_ETH, A_WETH
+from rotkehlchen.constants.misc import EXP18
+from rotkehlchen.db.evmtx import DBEvmTx
+from rotkehlchen.fval import FVal
+from rotkehlchen.types import ChainID, EvmTransaction, Location, Timestamp, deserialize_evm_tx_hash
+from rotkehlchen.utils.hexbytes import hexstring_to_bytes
+
+
+@pytest.mark.parametrize('ethereum_accounts', [['0x20A1CF262Cd3A42a50D226fD728104119e6fD0a1']])  # noqa: E501
+def test_v2_swap(database, ethereum_manager, eth_transactions):
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:19]: can you change the name to `test_balancer_v2_swap`?

File: rotkehlchen/tests/unit/decoders/test_balancer.py [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+            balance=Balance(amount=FVal('0.00393701451')),
+            location_label='0x20A1CF262Cd3A42a50D226fD728104119e6fD0a1',
+            notes='Burned 0.00393701451 ETH for gas',
+            counterparty='gas',
+            identifier=None,
+            extra_data=None,
+        ), HistoryBaseEntry(
+            event_identifier=evmhash,
+            sequence_index=1,
+            timestamp=1646375440000,
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_balancer.py:111]: In the test file (all the instances of `HistoryBaseEntry` can you fix the type here? Should be `TimesatmpMS`

**Review by yabirgb — COMMENTED** [review]
Thanks for rebasing it @nebolax ! It was really helpful. Just minor details and I noticed that is missing the accounting file. Can you also add it?

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+            action_items.append(action_item)
+
+        return action_items
+
+    def _decode_swap_creation(
+            self,
+            tx_log: EvmTxReceiptLog,
+            decoded_events: list[HistoryBaseEntry],
+            action_items: list[ActionItem],  # pylint: disable=unused-argument
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:108]: Can you write a docstring here xplaining what you do here and how the logic works? The following comment probably needs to go in the docstring.

Also explain what you do more. For example later I see

```python
if len(action_items) == 0 and from_token == self.weth.evm_address:
```

and I don't understand why/what that is. You need to explain code or write self-explainable code (where possible)

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    event.event_type = HistoryEventType.TRADE
+                    event.event_subtype = HistoryEventSubType.SPEND
+                    event.notes = f'Swap {event.balance.amount} {self.eth.symbol} in balancer V2 from {event.location_label}'  # noqa: E501
+                    event.counterparty = CPT_BALANCER_V2
+
+        return None, [action_item]
+
+    def decode_balancer_v2_event(
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:152]: This function seems superficial. Just add the negated condition on the top of `_decode_swap_creation` with a `return None, []` and you can use only that and save space.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        ):
+            return False
+
+        is_receive = asset.evm_address == action_items[-1].extra_data['to_token']
+
+        event.counterparty = CPT_BALANCER_V2
+        event.event_type = HistoryEventType.TRADE
+        if is_receive:
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.notes = f'Receive {event.balance.amount} {asset.symbol} in balancer V2 from {event.location_label}'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:205]: What exactly is this event? I am trying to understand the notes but I don't grok it.

`Receive 10 GNO in balancer V2 from 0x2b...20`

This sounds wrong. What did you get and from where? The information I need to see is that I got 10 GNO, the address on which I got them and for what reason. So

`Receive 10 GNO in 0x2b..01 from balancer v2 pool XYZ`

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+            action_items[-1].extra_data['to_token'] != tx_log.address or
+            event.balance.amount != to_amount
+        ):
+            return False
+
+        is_receive = asset.evm_address == action_items[-1].extra_data['to_token']
+
+        event.counterparty = CPT_BALANCER_V2
+        event.event_type = HistoryEventType.TRADE
+        if is_receive:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:203]: unecessary variable

```suggestion
        event.counterparty = CPT_BALANCER_V2
        event.event_type = HistoryEventType.TRADE
        if asset.evm_address == action_items[-1].extra_data['to_token']:  # is receive
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+
+    def _maybe_enrich_balancer_v1_events(
+        self,
+        token: 'EvmToken',
+        tx_log: EvmTxReceiptLog,  # pylint: disable=unused-argument
+        transaction: EvmTransaction,  # pylint: disable=unused-argument
+        event: HistoryBaseEntry,
+        action_items: list[ActionItem],  # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> bool:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:219]: docstring

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+        actions = self._decode_v1_pool_event(all_logs=all_logs)
+        if actions is None:
+            return False
+
+        for action in actions:
+            assert action.extra_data is not None
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:225]: What's this assertion for? I don't like seeing assertions in here. Is it to satisfy mypy only?

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+            if (
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:226]: holy shit that shit is fuuuuuuuuuuuuuuuuuuugly.

Is there no other way to make it easier to read? Perhaps a dict?

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    # otherwise this is the event that we have to edit
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Deposit {event.balance.amount} {token.symbol} in Balancer V1 pool'  # noqa: E501
+                    return True
+            if (
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:250]: This should have been elif no?
```suggestion
            elif (
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+            ):
+                if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    # otherwise this is the event that we have to edit
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Receive {event.balance.amount} {token.symbol} as deposit in Balancer V1 pool'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:238]: same as above. This notes make no sense.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+            ):
+                if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    # otherwise this is the event that we have to edit
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Receive {event.balance.amount} {token.symbol} after removing liquidity in Balancer V1 pool'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:262]:
```suggestion
                    event.notes = f'Receive {event.balance.amount} {token.symbol} after removing liquidity from a Balancer V1 pool'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+                    return True
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    # otherwise this is the event that we have to edit
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Return {event.balance.amount} {token.symbol} to Balancer V1 pool'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:272]:
```suggestion
                    event.notes = f'Return {event.balance.amount} {token.symbol} to a Balancer V1 pool'  # noqa: E501
```

**Review by LefterisJP — COMMENTED** [review]
Left some comments

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 14 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:250]: No since all `if` conditions above have a return statement

Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:250]: You are misreading the code, which shows why such ifs are bad. The ifs that have a return statement are indented. There is only one if above this, and it does not end with a return.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 12 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:225]: Yes. I'll change to an if is None then continue

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 9 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py]: I have added this information to the message. The balancer tokens have all the same name symbol so is not easy to obtain that information but everything else can be done

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 13 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:230]: I've tried to simplify a bit the logic, added more comments explaining things and removed the assertions

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 7 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:109]: I have added information to the docstring and in comments to clarify

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+                asset=self.eth,
+                amount=ZERO,
+                to_event_type=None,
+                to_event_subtype=None,
+                to_counterparty=CPT_BALANCER_V1,
+                to_notes=None,
+                extra_data={
+                    'ds_address': ds_address,
+                    'token_address': token_address,
+                    'amount': amount,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:94]: why keep the amount here? The `ActionItem` has an amount field.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+                from_event_subtype=HistoryEventSubType.NONE,
+                asset=self.eth,
+                amount=ZERO,
+                to_event_type=None,
+                to_event_subtype=None,
+                to_counterparty=CPT_BALANCER_V1,
+                to_notes=None,
+                extra_data={
+                    'ds_address': ds_address,
+                    'token_address': token_address,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:93]: Why not keep it at the `asset` of the `ActionItem`? Do you specifically need an address or an asset?

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: list[ActionItem],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        """
+        Decode swap in balancer V2. At the beggining of the transaction a SWAP event is created
+        with the information of the tokens and amounts and later some transfers are executed.
+
+        We need to detect this swap event and then match the transferred amounts with the ones
+        in the swap event. A sepcial case is the swap of ETH that is wrapped before being sent.
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:115]:
```suggestion
        in the swap event. A special case is the swap of ETH that is wrapped before being sent.
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+            asset=self.eth,
+            amount=ZERO,
+            to_event_type=None,
+            to_event_subtype=None,
+            to_counterparty=CPT_BALANCER_V2,
+            to_notes=None,
+            extra_data={
+                'from_token': from_token,
+                'to_token': to_token,
+                'amount_in': amount_in,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:141]: you can use asset and amount which are already part of the `ActionItem` for the _in and then extra_data for the _out, but properly document it. Otherwise you end up wasting memory by not using the ActionItem's field and just using the extra data.

Perhaps if this is done in many decoders we may need to rethink how some ActionItems work

If it becomes too complicated you can leave it like this but ... hm ... damn it's ugly to have so many unused fields and use extra data

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+            self,
+            token: 'EvmToken',  # pylint: disable=unused-argument
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            event: HistoryBaseEntry,
+            action_items: list[ActionItem],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> bool:
+        """
+        Enrich tranfer transactions to address for swaps in balancer v2 protocol.
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:178]:
```suggestion
        Enrich tranfer transactions to account for swaps in balancer v2 protocol.
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+            action_items[-1].extra_data['to_token'] != tx_log.address or
+            event.balance.amount != to_amount
+        ):
+            return False
+
+        event.counterparty = CPT_BALANCER_V2
+        event.event_type = HistoryEventType.TRADE
+        if asset.evm_address == action_items[-1].extra_data['to_token']:
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.notes = f'Receive {event.balance.amount} {asset.symbol} in Balancer V2 from {event.location_label}'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:209]: How does this read. I don't get it. `Receive 10 GNO in Balancer V2 from 0xmyaddress`?

This sounds wrong.

Shouldnt' it be `Receive 10 GNO from Balancer v2` or ... if anything more have a pool or something?

Also I see you use capitalized V2. Don't do it. Change it everywhere. It should be v2.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+        event: HistoryBaseEntry,
+        action_items: list[ActionItem],  # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> bool:
+        """
+        Enrich balancer v1 transfer to read pool events of:
+        - Depositing in the pool
+        - Withdrawing from the pool
+
+        In balancer V1 pools are managed using a DSProxy so the account doesn't interact
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:229]:
```suggestion
        In balancer v1 pools are managed using a DSProxy so the account doesn't interact
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+
+            if action.extra_data['type'] == BalancerV1EventTypes.JOIN:
+                if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Receive {event.balance.amount} {token.symbol} at {event.location_label} from Balancer V1 pool'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:253]:
```suggestion
                    event.notes = f'Receive {event.balance.amount} {token.symbol} at {event.location_label} from a Balancer v1 pool'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+                    return True
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Deposit {event.balance.amount} {token.symbol} in Balancer V1 pool from {event.location_label}'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:262]:
```suggestion
                    event.notes = f'Deposit {event.balance.amount} {token.symbol} to a Balancer v1 pool from {event.location_label}'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+                    return True
+            elif action.extra_data['type'] == BalancerV1EventTypes.EXIT:
+                if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Receive {event.balance.amount} {token.symbol} in {event.location_label} after removing liquidity from a Balancer V1 pool'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:272]: You seem to use different prepositions across different notes. Please be consistent.

```suggestion
                    event.notes = f'Receive {event.balance.amount} {token.symbol} at {event.location_label} after removing liquidity from a Balancer v1 pool'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+                    return True
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ):
+                    event.event_type = HistoryEventType.STAKING
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.counterparty = CPT_BALANCER_V1
+                    event.notes = f'Return {event.balance.amount} {token.symbol} to a Balancer V1 pool from {event.location_label}'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:281]:
```suggestion
                    event.notes = f'Return {event.balance.amount} {token.symbol} to a Balancer v1 pool from {event.location_label}'  # noqa: E501
```

**Review by LefterisJP — COMMENTED** [review]
Left more comments

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 18 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:93]: I have moved it. Due to this change I had to add a new type in the action items

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.base_tools = base_tools
+        self.eth = A_ETH.resolve_to_crypto_asset()
+        self.weth = A_WETH.resolve_to_evm_token()
+
+    def _decode_v1_pool_event(self, all_logs: list[EvmTxReceiptLog]) -> Optional[list[ActionItem]]:
+        """Read the list of logs in search for a Balancer V1 event and return the information
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:54]:
```suggestion
        """Read the list of logs in search for a Balancer v1 event and return the information
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 19 above)
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:92]: you have not answered this

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 18 above)
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:93]: moved what? This is still here.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 21 above)
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:145]: no answer here or resolution by you but I see you use what I suggested, so will resolve this.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+            to_counterparty=CPT_BALANCER_V2,
+            to_notes=None,
+            extra_data={
+                'from_token': from_token_address,
+                'amount_in': amount_in,
+            },
+        )
+
+        # When ETH is swapped it is wrapped to WETH and the ETH transfer happens before the SWAP
+        # event, we need to detect it we haven't done it yet.
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:150]:
```suggestion
        # event. We need to detect it if we haven't done it yet.
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 31]
```diff
@@ -0,0 +1,296 @@
...
+        - WrongAssetType
+        """
+        if action_items is None or len(action_items) == 0 or transaction.to_address != VAULT_ADDRESS:  # noqa: E501
+            return False
+
+        if action_items[-1].extra_data is None:
+            return False
+
+        asset = event.asset.resolve_to_evm_token()
+
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:193]:
```suggestion
        asset = event.asset.resolve_to_evm_token()
```

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 32]
```diff
@@ -0,0 +1,296 @@
...
+        if (
+            isinstance(action_items[-1].asset, EvmToken) is False or
+            action_items[-1].asset.evm_address != tx_log.address or  # type: ignore[attr-defined]  # noqa: E501 It fails to detect the previous comprobation
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:196]:
```suggestion
            action_items[-1].asset.evm_address != tx_log.address or  # type: ignore[attr-defined]  # noqa: E501 It fails to detect the previous MUMBO JUMBO
```

> comprobation

I googled this. It's a real word but I bet you not even the late queen's grandma knows what it means. If you explain something in a comment use plain English. I google it and I still don't get what you want to say. When you write a comment you are ELI5ing to a kid.

Also if it's too long you can always write it above the condition where you type ignored.

You can check with me in discord about the explanation.

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 33]
```diff
@@ -0,0 +1,296 @@
...
+            action_items[-1].amount != event.balance.amount
+        ):
+            return False
+
+        event.counterparty = CPT_BALANCER_V2
+        event.event_type = HistoryEventType.TRADE
+        if asset == event.asset:
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.notes = f'Receive {event.balance.amount} {asset.symbol} in Balancer v2 from {event.location_label}'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:205]: This has remained unchanged and my question has not been answered. Please try to appreciate the time I am putting in the reviews **and read them**

https://github.com/rotki/rotki/pull/5197#discussion_r1051653578

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py [hunk 34]
```diff
@@ -0,0 +1,296 @@
...
+
+    def _maybe_enrich_balancer_v1_events(
+        self,
+        token: 'EvmToken',
+        tx_log: EvmTxReceiptLog,  # pylint: disable=unused-argument
+        transaction: EvmTransaction,  # pylint: disable=unused-argument
+        event: HistoryBaseEntry,
+        action_items: list[ActionItem],  # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> bool:
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/balancer/decoder.py:219]: This entire function declaration needs indentation

**Review by LefterisJP — COMMENTED** [review]
comments

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 23 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py]: Yes, as discussed changed it to `Receive 10 GNO from Balancer v2`

File: rotkehlchen/chain/ethereum/modules/balancer/decoder.py (hunk 19 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/balancer/decoder.py:92]: I have changed this function to return a dictionary instead. I was using actions before because of how I was working with the events but saw that it could be done easier just calling the logic in the `_decode_v1_pool_event` function. It doesn't need and it shouldn't be an action

**Review by LefterisJP — APPROVED** [review]
LGTM.

1. The failed test is avalanche
2. The single comment can be done whenever.

When you do the comment also please add a changelog entry for proper decoding of balancer transactions.
